core/src/main/java/org/infinispan/container/impl/InternalDataContainer.java

```diff
@@ -146,6 +155,45 @@ default int size(IntSet segments) {
     */
    Iterator<InternalCacheEntry<K, V>> iteratorIncludingExpired(IntSet segments);
 
+   /**
+    * Performs the given action for each element of the container that maps to the given set of segments
+    * until all elements have been processed or the action throws an exception.  Unless otherwise specified by the
+    * implementing class, actions are performed in the order of iteration (if an iteration order is specified).
+    * Exceptions thrown by the action are relayed to the caller.
+    *
+    * @implSpec
+    * <p>The default implementation behaves as if:
+    * <pre>{@code
+    *   for (Iterator<InternalCacheEntry<K, V>> iter = iterator(segments) ; iter.hasNext() ; ) {
+    *      InternalCacheEntry<K, V> ice = iter.next();
+    *      action.accept(ice);
+    *   }
+    * }</pre>
+    *
+    * @param action The action to be performed for each element
+    * @throws NullPointerException if the specified action is null
+    */
+   default void forEach(IntSet segments, Consumer<? super InternalCacheEntry<K, V>> action) {
+      for (Iterator<InternalCacheEntry<K, V>> iter = iterator(segments) ; iter.hasNext() ; ) {
+         InternalCacheEntry<K, V> ice = iter.next();
+         action.accept(ice);
+      }
+   }
+
+   /**
+    * Performs the given action for each element of the container, even if expired. This method should be preferred when
+    * user wants to perform an operation across all entries that depends on the segment it maps to.
+    * @param action The action to be performed for each element
+    * @throws NullPointerException if the specified action is null
+    */
+   void forEachIncludingExpired(ObjIntConsumer<? super InternalCacheEntry<K, V>> action);
+
+   /**
+    * Sets what segments this data container should be using. This can contain already started segments.
+    * @param segments segments to add
```

Comment:
I think we need to settle on a single word to describe these segments :)

That's why I suggested defining the notion of owned/associated/started segments in the interface javadoc, there we can also say that impls are free to reject writes to non-owned segments. And if you prefer not using `clear(segments)` to remove the L1 entries, you can also say that any change in owned segments is allowed to clear the non-owned segments.

Reply:
Updating docs to be more clear on this.